### PR TITLE
Allow installing a second agent on the same machine for development

### PR DESCRIPTION
**Pull request by cmacknz**
- Closes https://github.com/elastic/ingest-dev/issues/3260
- Depends https://github.com/elastic/elastic-agent/pull/4834

Allows installing or running a second Elastic Agent on the same machine for development using the `--develop` argument. The core problem this PR aims to solve is allowing agent developers to install a second agent on their machines when there is also an InfoSec managed agent installed. Once installed, the second agent is available with the `elastic-development-agent` command. I put `development` in the middle instead of at the end because it makes tab based autocompletion in the shell easier.

I have manually tested that the development agent can enroll with Fleet and upgrade. The changes needed for a second agent are isolated to install time and interacting with the daemon. Features like upgrades and enrolling in Fleet require no changes because they are contained into the agent's installation directory. For this reason I have only added integration tests to ensure installing a second agent alongside an existing one works. We shouldn't need to test anything else as there are no features that reach outside the agent installation directory.

I sanity checked what happens if you install Defend a second time on the same machine to make sure nothing explodes, and it fails immediately in a nice place before the second install is even attempted:

```sh
sudo elastic-development-agent status
┌─ fleet
│  └─ status: (HEALTHY) Connected
└─ elastic-agent
├─ status: (DEGRADED) 1 or more components/units in a failed state
└─ endpoint-default
    ├─ status: (FAILED) failed to start connection info service endpoint: failed to start connection credentials listener: listen tcp 127.0.0.1:6788: bind: address already in use
    ├─ endpoint-default
    │  └─ status: (FAILED) failed to start connection info service endpoint: failed to start connection credentials listener: listen tcp 127.0.0.1:6788: bind: address already in use
    └─ endpoint-default-53f5058f-8a9f-460d-b36c-5d4db6131658
        └─ status: (FAILED) failed to start connection info service endpoint: failed to start connection credentials listener: listen tcp 127.0.0.1:6788: bind: address already in use
```

Reproducing the added README section for instructions on how to install and test:

### Development Installations

> :warning: Development installations are not officially supported and are intended for Elastic Agent developers.

If you are an Elastic employee, you already have an Information Security managed Elastic Agent installed on your machine for endpoint protection. This prevents you from installing the Elastic Agent a second time for development without using a VM or Docker container. To eliminate this point of friction, Elastic Agent has a development mode that permits installing the Elastic Agent on your machine a second time:

```sh
# All other arguments to the install command are still supported when --develop is specified.
sudo ./elastic-agent install --develop
# The run command also supports the --develop option to allow running without installing when there is another agent on the machine.
./elastic-agent run -e --develop
```

Using the `--develop` option will install the agent in an isolated `Agent-Development` agent directory in the chosen base path.

Development agents enrolled in Fleet will have the `Development` tag added automatically. Using the default base path on MacOS you will see:

```sh
sudo ls /Library/Elastic/
Agent
Agent-Development
```

The `elastic-agent` command in the shell is replaced with `elastic-development-agent` to interact with the development agent:

```sh
# For a privileged agent
sudo elastic-development-agent status
# For an unprivileged agent
sudo -u elastic-agent-user elastic-development-agent status
```

The primary restriction of `--develop` installations is that you cannot run Elastic Defend a second time on the same machine. Attempting toinstall Defend twice will fail with resource conflicts. All other integrations should be usable provided conflicting configurations are changed ahead of time. For example two agents cannot bind to the same `agent.monitoring.http.port` to expose their monitoring servers

### CODE REVIEW

File: internal/pkg/agent/application/upgrade/rollback.go [hunk 1]
```diff
@@ -146,20 +146,20 @@ func InvokeWatcher(log *logger.Logger, agentExecutable string) (*exec.Cmd, error
...
 	log.Infow("Starting upgrade watcher", "path", cmd.Path, "args", cmd.Args, "env", cmd.Env, "dir", cmd.Dir)
 	if err := cmd.Start(); err != nil {
 		return nil, fmt.Errorf("failed to start Upgrade Watcher: %w", err)
 	}
 
 	upgradeWatcherPID := cmd.Process.Pid
 	agentPID := os.Getpid()
+
+	go func() {
+		if err := cmd.Wait(); err != nil {
```
Comment (cmacknz) [inline internal/pkg/agent/application/upgrade/rollback.go:158]: I am going to separate this change into its own PR.

**Comment by cmacknz** [comment]
Pulling out of draft. The changes needed were more straightforward than I anticipated.

File: internal/pkg/agent/application/upgrade/rollback.go (hunk 1 above)
Comment (cmacknz) [reply internal/pkg/agent/application/upgrade/rollback.go:158]: https://github.com/elastic/elastic-agent/pull/4834 which has some implications for the correctness of the watcher on Unix.

**Comment by cmacknz** [comment]
buildkite test it

**Comment by blakerouse** [comment]
I wonder if this is the correct way we should handle this with the additional `--develop` flag. Would it be better to just support multiple Elastic Agent's on a single host?

I know there have been requests to have multiple Elastic Agent's running on the same host. I think it might be better to just support multiple at a same time in a clean way.

**Comment by cmacknz** [comment]
>  I wonder if this is the correct way we should handle this with the additional --develop flag. Would it be better to just support multiple Elastic Agent's on a single host?

I do like `--develop` as a shorthand for installing into a well know location (and maybe it lets us doing some special additional things later), but yes it wouldn't be much additional work to support arbitrary numbers of agents by introducing an installation namespace concept.

I'm not sure we want to start officially recommending this, but I can refactor a bit to support it easily.

**Comment by blakerouse** [comment]
> > I wonder if this is the correct way we should handle this with the additional --develop flag. Would it be better to just support multiple Elastic Agent's on a single host?
> 
> I do like `--develop` as a shorthand for installing into a well know location (and maybe it lets us doing some special additional things later), but yes it wouldn't be much additional work to support arbitrary numbers of agents by introducing an installation namespace concept.
> 
> I'm not sure we want to start officially recommending this, but I can refactor a bit to support it easily.

I defiantly know there is a use-case where you have a divide between teams in organizations and they each want to run an Elastic Agent on the same host. Be interested to get @nimarezainia opinion on it.

**Comment by cmacknz** [comment]
OK I refactored this to add the concept of an installation namespace, that can be any arbitrary value, and would allow the installation of any number of agents at once.

However I haven't exposed the ability to install into arbitrary namespaces, I've restricted it to a single well known `Development` namespace via the `--develop` option.

The current thinking is we don't actually want to officially support multiple agents on the same machine yet, as it is often a crutch for Fleet features that will be delivered soon (e.g. RBAC) and making it work nicely in the Fleet UI would require some extra polish. If we change our minds on this, all we need to do now is add a `--namespace` argument to the install command that accepts arbitrary values. I also don't think we want to deal with the consequences of internal users installing 5 agents on the same machine yet before we get a sense that 2 works reasonably.

I double checked that it worked by hard coding in the arbitrary namespace `foobar` with the following in install.go:

```
	isDevelopmentMode, _ := cmd.Flags().GetBool(flagInstallDevelopment)
	if isDevelopmentMode {
		fmt.Fprintln(streams.Out, "Development installation mode enabled; this is an experimental and currently unsupported feature.")
		paths.SetInstallNamespace("foobar")
	}
```

This gives me an `elastic-foobar-agent` shell wrapper with the agent in the `/Library/Elastic/Agent-foobar` directory on my machine as expected.

File: internal/pkg/agent/configuration/grpc.go [hunk 2]
```diff
@@ -18,11 +20,21 @@ type GRPCConfig struct {
 
 // DefaultGRPCConfig creates a default server configuration.
 func DefaultGRPCConfig() *GRPCConfig {
+	// When in an installation namepspace, bind to port zero to select a random free port to avoid
```
Comment (blakerouse) [inline internal/pkg/agent/configuration/grpc.go:23]:
```suggestion
	// When in an installation namespace, bind to port zero to select a random free port to avoid
```

**Review by blakerouse — COMMENTED** [review]
Really like this! Just a few questions.

File: internal/pkg/agent/application/paths/common_namespace.go [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+// ServiceName returns the service name accounting for any namespace.
+func ServiceName() string {
+	namespace := InstallNamespace()
+	if namespace == "" {
+		return serviceName
+	}
+
+	// Use strings.Replace() to avoid having to sanitize format specifiers in the namespace itself.
+	return strings.Replace(serviceNameNamespaceFmt, "%s", namespace, 1)
```
[the remark this answers is not in the document]
Comment (cmacknz) [reply internal/pkg/agent/application/paths/common_namespace.go:96]: For some reason I had it in my head that I had to worry about escaping things using Sprintf but rewriting it that makes no sense since the format string is a constant.

Updated.

File: internal/pkg/agent/application/paths/common_namespace.go [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+	if RunningInstalled() {
+		return parseNamespaceFromDir(filepath.Base(Top()))
+	}
+
+	return ""
+}
+
+func parseNamespaceFromDir(dir string) string {
+	parts := strings.SplitAfterN(dir, "-", 2)
+	if len(parts) <= 1 {
```
[the remark this answers is not in the document]
Comment (cmacknz) [reply internal/pkg/agent/application/paths/common_namespace.go:55]: Yes good idea.

File: internal/pkg/agent/application/paths/common_namespace.go [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+// directory name, since a unique directory name is required to avoid collisions between installed
+// agents in the same base path. Before installation, the installation namespace must be configured
+// using SetInstallNamespace().
+func InstallNamespace() string {
+	if installNamespace != "" {
+		return installNamespace
+	}
+
+	if RunningInstalled() {
+		return parseNamespaceFromDir(filepath.Base(Top()))
```
[the remark this answers is not in the document]
Comment (cmacknz) [reply internal/pkg/agent/application/paths/common_namespace.go:47]: Good idea and trivial, done.

File: internal/pkg/agent/cmd/install.go [hunk 6]
```diff
@@ -49,6 +50,8 @@ would like the Agent to operate.
 	cmd.Flags().String(flagInstallBasePath, paths.DefaultBasePath, "The path where the Elastic Agent will be installed. It must be an absolute path.")
 	cmd.Flags().Bool(flagInstallUnprivileged, false, "Installed Elastic Agent will create an 'elastic-agent' user and run as that user. (experimental)")
 	_ = cmd.Flags().MarkHidden(flagInstallUnprivileged) // Hidden until fully supported
+	cmd.Flags().Bool(flagInstallDevelopment, false, "Install an isolated Elastic Agent for development. Allows a non-development agent to be installed already. (experimental)")
+	_ = cmd.Flags().MarkHidden(flagInstallDevelopment) // For internal use only.
```
[the remark this answers is not in the document]
Comment (cmacknz) [reply internal/pkg/agent/cmd/install.go:54]: "Two agents ought to be enough for anybody."

If I expose `--namespace` then I have to go off and test it, and I want to keep the scope of what this lets us do limited to what we need while letting us expand it as needed in the future.

I'm not strictly opposed to exposing this and hiding it, but I'd prefer it be done as a follow up. Now that ON week is over I have a lot less time to consistently dedicate to this and it's useful in its current form.

File: .golangci.yml [hunk 7]
```diff
@@ -124,7 +124,7 @@ linters-settings:
 
   nolintlint:
     # Enable to ensure that nolint directives are all used. Default is true.
-    allow-unused: false
+    allow-unused: true
```
Comment (cmacknz) [inline .golangci.yml:127]: I did this because I had platform dependent lint warnings and this seems like the only wait to deal with that reasonably.

File: internal/pkg/agent/cmd/install.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (blakerouse) [reply internal/pkg/agent/cmd/install.go:54]: I agree that two agents ought to be enough for anybody. I just don't agree that forcing the name to `development` as that solution. I can see a case in production environments that would require two Elastic Agent's and I know they would not want one of them to be named `development`.

Not a blocker, can possibly expose `--namespace` in a follow up, if we end up supporting for real deployments.

[the remark this answers is not in the document]
Comment (cmacknz) [reply internal/pkg/agent/cmd/install.go:58]: I added `--namespace`, but did not document it. In the integration tests I alternately use --namespace and --develop.

I like --develop as a shorthand, I can also mirror it to the run command where --namespace makes no sense but we still need some special setup to avoid needing extra configuration when running two agents on the same machine.

**Comment by alexsapran** [comment]
that would be great, as this is something that would be helpful with the work I am doing.

Having the ability to just download an elastic-agent and just run it locally without installing it.

**Review by blakerouse — APPROVED** [review]
Looks good thanks for the fixes and the exposing of `--namespace` optional as a command line option.

**Comment by cmacknz** [comment]
> [31.3% Coverage on New Code](https://sonar.elastic.dev/component_measures?id=elastic_elastic-agent_AYluowg0xMq8P7b4moiZ&pullRequest=4822&metric=new_coverage&view=list) (required ≥ 40%)

I am going to force merge past this, most of the uncovered code is from the conversion of variables into functions and the uncovered code is tested in integration tests.
